[FRONTEND] Add search product page

MehdiSaffar opened the pull request:
## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Description

Product search page with filter, sort and pagination

- 3 subcomponents (top: search bar, left: search filters, right: search results)
- deterministic: uses query params, every submit changes them (for example `/search/product?brand=adidas&category=home-and-garden&priceRange=19,58&query=smart%20watch&rating=3.5&sortBy=price&subcategory=pet-supplies&vendor=mosaic`) easily parsable with [query-string](https://www.npmjs.com/package/query-string) package
- left panel has:
  - Category
  - Subcategory (changes dynamically based on Category)
  - Vendor
  - Brand
  - Price range
  - Rating
  - Sort by
    - Best sellers
    - Newest arrivals
    - Price
    - Average customer review (rating)
    - Number of comments
- right panel has:
  - fluid product grid (with flex-wrap)
  - pagination at the bottom (won't show if single page result)
- top is search bar: when pressed enter, left panel will result (expected result as in many websites)
- left panel slides down with the view when scrolling 

# Screenshots

## Main view 
![image](https://user-images.githubusercontent.com/13124263/99654895-0c87e800-2a6c-11eb-90ec-577a349fcabf.png)

## Side panel follows you when scrolling
Also showing pagination 
![image](https://user-images.githubusercontent.com/13124263/99654927-19a4d700-2a6c-11eb-9dec-2d1ac062c2f4.png)

# Test results

No tests.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have checked my code and corrected any misspellings
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# Potential problems/Important remarks

- pagination is done on server-side, only information such as current page, count per page required for correct display
- refresh on pagination change not implemented yet (EDIT: done)
- needs to be tested with some better mock json api meanwhile
- product cards need to be made smaller
- missing reset filters (EDIT: done)
- search bar might be changed to have a dropdown of Vendor vs Product instead categories
- left panel's rating and price range a bit hard to see with current background (EDIT: done)
- need to think of whether integrate vendor search in same react component or not, and whether seperate frontend endpoints like `'/search/vendor/` and `/search/product/` or unified. (EDIT: done, separate url but depends on `type` prop)

MehdiSaffar committed:
most search panels fields rendering

MehdiSaffar committed:
sticky search panel, showing a few fake products with wrap

MehdiSaffar committed:
adjust gaps

MehdiSaffar committed:
add pagination and props to SearchResults

MehdiSaffar committed:
refactor to seperate files

MehdiSaffar committed:
using query params properly

MehdiSaffar committed:
made search page more deterministic on query params

MehdiSaffar committed:
show spinning even for first time

MehdiSaffar committed:
Merge branch 'master' into F-search-page

MehdiSaffar committed:
Merge branch 'master' into F-search-page

MehdiSaffar committed:
sticky top: 20px

MehdiSaffar committed:
Properly doing query params

- Refactor query params handling
- Spacing pagination from product grid
- Add all pagination props

MehdiSaffar committed:
add reset to initial values

MehdiSaffar committed:
padding and bg color to side panel

MehdiSaffar committed:
changed url for product search

ozdenizdolu commented on ecommerce/frontend/src/components/search/SearchSidePanel.js:
```diff
@@ -0,0 +1,123 @@
+import { Button, Form, Rate, Select, Slider } from 'antd'
+import { useState } from 'react'
+
+import { brands, categories, productSortBy, subcategories, vendors } from '../../utils'
```
thanks for adding these utils, they will be helpful to us as well

ozdenizdolu reviewed (approved):
I approved the changes. Couldn't have done it better myself.

Commit:
Small fixes on query parsing and refreshing url

MehdiSaffar committed:
add better title in search

MehdiSaffar commented:
Added title that depends on query or subcategory or category basically something like this:

```js
        const prefix = query ?? subcategory ?? category ?? 'All'

        return `${prefix} ${type}s - Getflix`
```

abdullahyldz commented on ecommerce/frontend/src/components/search/SearchSidePanel.js:
```diff
@@ -0,0 +1,123 @@
+                    ))}
+                </Select>
+            </Form.Item>
+            <Form.Item name="priceRange" label="Price range">
+                <Slider range />
+            </Form.Item>
+            <Form.Item name="rating" label="Rating">
+                <Rate allowHalf allowClear />
+            </Form.Item>
```
Although original Slider in the antd page is colored blue, my browser renders it as white. 
Also, Rate component is very hard to see. Lets give default value as 5 stars to make it more visible. 

![image](https://user-images.githubusercontent.com/44057640/99707166-2564bd80-2aad-11eb-8015-8a5e08680858.png)
![image](https://user-images.githubusercontent.com/44057640/99707219-331a4300-2aad-11eb-9365-8efcfa612be7.png)

abdullahyldz replied:
Also, Slider default value should be [0 to Max](currently [0-0] by default since we include every item.

abdullahyldz replied:
Oh, you are getting the default values from the query-string parse method. Then, they start with empty default values.

abdullahyldz replied:
I am sorry, I was a couple commit behind you. You added some background and all. It looks better with that purple color

furkan-varol replied:
I agree with Abdullah about the Rating component. The default should be 5 stars because it just looks more appealing.

abdullahyldz reviewed (approved):
Very good job, indeed.

furkan-varol commented:
I think the dropdown for category selection could be wider. Not each category is fully shown as it is.
![image](https://user-images.githubusercontent.com/44326869/99796960-6b6a6180-2b3f-11eb-8005-808519110939.png)

furkan-varol commented on ecommerce/frontend/src/components/pages/SearchPage.js:
```diff
@@ -0,0 +1,150 @@
+import qs from 'query-string'
+import { useEffect, useState } from 'react'
+import { subcategories, categories, sleep } from '../../utils'
+import Axios from 'axios'
```
Unnecessary import.

MehdiSaffar committed:
Multiple changes (read desc)

- products/vendors dropdown in search input instead of categories
- change in the query params object to adapt to search input
- remove throw error in SearchPage
- better color for unfilled stars
- skipNull and skipEmptyString in query stringify
- add initialValues to SearchInput
- replace price range with max price which can be null
- refactor Content routing make it more correct

MehdiSaffar committed:
make dropdown borderless in search input

MehdiSaffar committed:
fix color on active half star

MehdiSaffar commented:
I have incorporated all changes.
